Adding tidal profile generator

malcolmmoncheur opened the pull request.

malcolmmoncheur committed:
added dummy functions for:
 get_load_tidal_from_upload
 extract_tidal_constituents
 extrapolate_tidal_epoch
 generate_tidal_profiles

malcolmmoncheur committed:
modified dummy functions:
 get_load_tidal_from_upload
 extract_tidal_constituents
 extrapolate_tidal_epoch
 generate_tidal_profiles

malcolmmoncheur committed:
modified dummy functions:
 get_tidal_data_from_upload
 extract_tidal_constituents
 extrapolate_tidal_epoch
 generate_tidal_profiles

modified if __name__ == "__main__":

malcolmmoncheur committed:
updated generate_tidal_profile.py

The code runs in debug mode

malcolmmoncheur committed:
updated generate_tidal_profile.py

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+
+        if validate:
+            # List of initialized parameters to validate
+            args_dict = {'latitude': self.latitude,
```
We should add the new params here and add validation checks for them, in a follow-on task.

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+    def extrapolate_tidal_epoch(self):
+        """Extract tidal constituents from 8760 of tidal current data and extrapolate 19-year tidal epoch"""
+
+        # could modify this in the future to extrapolate any amount of data
```
It can be helpful to add "TODO" to the beginning of inline comments. Then those lines get flagged by PyCharm and other IDEs.

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+    def get_dc_to_ac(self):
+        """ Returns the dc to ac ratio. """
+
+        # Should we keep this? I don't think this matters
```
We probably don't need this, esp since we don't have the 'inverter_output' param defined.

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+    def tidal_checks(self):
+        """ Several checks to  make sure the tidal profiles look OK. """
+
+        ## Do we need this? I left this as-is from the solar.
```
Let's keep this in for now.

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+
+    def crop_timeline(self, num_seconds, validate=True):
+        """ Used to crop the profiles to a period of less than 1 day
+        ## Is this necessary? I kept this the same from 'generate_solar_profile'
```
I don't think we need this, esp because trial length is now specified in hours.

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+    def get_turbine_losses(self):
+        """ Returns the system losses. """
+
+        # I don't think this is necessary either, since losses are
```
Agreed.

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+
+    for index, row in tidal_profile.iterrows():
+        u = row['v']
+        if u > cut_in_velocity and u < cut_out_velocity:
```
Should it be >= and <=?

malcolmmoncheur replied:
Yes, I think that's right. Good catch

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+        elif u < cut_in_velocity:
+            dc_power.at[index, 'power'] = 0
+        elif u > cut_out_velocity:
+            dc_power.at[index, 'power'] = rated_power / (rated_power * turbine_number)
```
Is this equation correct?

malcolmmoncheur replied:
If tidal turbines operate similarly to wind turbines, than my assumption is that the the capacity of a specific turbine (e.g. 50kW) is equal to the peak production of that turbine. So if we we're above the cut-out speed, the turbine is limited to operating at max production. Assuming a 50kW turbine, if above cut-out speed, the power equation outputs that the turbine is producing a continuous 50kW of power. However, I am trying to calculate power output per 1kW of tidal, which is why the output is divided by the rated power of the turbine times how many turbines there are

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+
+
+    # Calculate AC power
+    ac_power = dc_power['power'] * inverter_efficiency
```
Divide inverter_efficiency by 100?

malcolmmoncheur replied:
Good catch. Instead of doing that, I changed the inverter efficiency to be a decimal - that way it matches the other input formats.

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+    for index, row in tidal_profile.iterrows():
+        u = row['v']
+        if u > cut_in_velocity and u < cut_out_velocity:
+            dc_power.at[
```
Shouldn't the power produced be multiplied by the number of turbines not divided?

malcolmmoncheur replied:
See my note above. I'm calculating the power generation per 1kW of tidal, so the total power produced is divided by the rated power of the turbine... although I'm now noticing that it should also be multiplied by the number of turbines - so that should cancel out dividing by the number of turbines,

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -0,0 +1,438 @@
+    ac_power[ac_power < 0] = 0
+
+    # Divide by the total DC power to get the power (in kW) per 1kW of tidal
+    return ac_power / rated_power
```
Should this also be divided by # turbines?

malcolmmoncheur replied:
I actually don't think this line is required anymore, since the dc power calculation already normalizes to 1kW of tidal

SarahFNewman reviewed (changes requested):
Great job! Some minor comments and then questions around the power calculation. I suggest looking at PEP8 guidance on blank lines as well: https://peps.python.org/pep-0008/#blank-lines

malcolmmoncheur commented:
I think I made all the changes you suggested. Would love to talk to you about how we want to present the normalized turbine output. And thank you for the blank lines guidance - I had forgotten to clean things up, so I appreciate it.

malcolmmoncheur committed:
- cleaned up blank spaces
- updated power calculation
- made minor edits as requested

malcolmmoncheur committed:
- updated normalized power production calculation

SarahFNewman reviewed (approved):
Looks great!

malcolmmoncheur committed:
- added validation parameters to parameter_validation.csv
- slightly tweaked power production equation to account for turbines with multiple rotors

SarahFNewman commented on data/parameter_validation.csv:
```diff
@@ -65,9 +90,13 @@ num_seconds,int,1,86400,,,,,
 one_way_battery_efficiency,"int,float",0,1,,,,,
 one_way_inverter_efficiency,"int,float",0,1,,,,,
 outputs,dict,,,,,check_outputs,,Something is wrong with the aggregated system outputs.
+Parameter,Type,Min,Max,Enums,Size,Custom,Custom Arguments,Custom Message
```
We should keep this line at the top.

malcolmmoncheur replied:
Sorry about that - I accidentally had the header row get sorted alphabetically with the rest of the data. fixed

SarahFNewman commented on generate_tidal_profile.py:
```diff
@@ -305,14 +270,15 @@ def calc_tidal_prod(tidal_profile, latitude, longitude,
                      'turbine_losses': turbine_losses,
-                     'cut_in_velocity': 0.7,
-                     'cut_out_velocity': 2.6}
+                     'cut_in_velocity': 0.5,
```
These should be set by the input args and not hard-coded.

malcolmmoncheur replied:
Thanks - updated this

SarahFNewman commented on data/parameter_validation.csv:
```diff
@@ -1,23 +1,30 @@
 cld_hours,tuple,,,,2,check_cld_hours,,each element in cld_hours must be between 0 and 23
 component,microgrid_system.Component,,,,,,,
+connections,"int,float",0,100,,,,,
+cut_in_velocity,"int,float",0,,,,,,
```
Is there a possibility that "reasonable" values for cut-in and cut-out velocity would be different for tidal, wave, and wind turbines? If so, we may want to make this param more specific, i.e. "tidal_cut_in_velocity".

malcolmmoncheur replied:
Right now I only have the lower bound specified (0) for both. but you are correct. typically the minimum average wind velocity you would want is 5m/s. So I followed your advice and appended "tidal_" to the front of the cut in and cut out speeds

malcolmmoncheur replied:
I also went ahead and added "tidal_" ahead of other values that could be used in a potential wave or wind module (inverter efficiency, number of turbines, number of rotors, etc)

SarahFNewman commented on data/parameter_validation.csv:
```diff
@@ -77,69 +106,50 @@ power_profile,pandas.core.series.Series,,,,,check_power_profile,,"The power prof
 rated_power,"int,float",0,,,,,,
 read_from_file,bool,,,,,,,
 resilience_percent,"int,float",0,100,,,,,
+rotor_length,"int,float",0,50,,,,,
```
Same comment as for cut-in speed. If we think that the range could be different for other technologies that we may be adding soon (not sure whether or not this is the case), this should be tidal-specific. Same for rotor_number.

SarahFNewman reviewed (changes requested):
Very minor comments on the validation stuff, but I'm having trouble running the code. A few notes on that:
- Can you add utide to the requirements file?
- Is there sample data that is needed to run the main function at the bottom of generate_tidal_profile?
- I'm getting an error that stems from the min/max values from the validation file being read in as strings instead of floats/ints. I suspect that when you put the header row back at the top, this should fix it. Are you seeing the same thing?

malcolmmoncheur commented:
- I added utide to the requirements file
- yes, running the main function at the bottom of tidal_profile_generator requires "portangeles_2015_9m.csv". I forgot to add it to git - it's been added now
- I am not getting that error. let me know if you keep having it now that I've fixed the csv file header row

malcolmmoncheur committed:
- made changes based on Sarah's feedback:
- added sample tidal data
- appended "tidal_" in front of all tidal variables
- updated parameter_validation.csv

malcolmmoncheur committed:
- made changes based on Sarah's feedback:
- added sample tidal data
- appended "tidal_" in front of all tidal variables
- updated parameter_validation.csv

malcolmmoncheur committed:
updated requirements.txt

SarahFNewman reviewed (changes requested):
Looks great. One more tiny comment. The directory "tidal_profiles" needs to already exist for the code to run. To do that, create an empty file called ".gitkeep" within the directory and then add that file to git and commit. That will allow the empty directory to be included in the repo.

malcolmmoncheur committed:
added tidal_profiles folder to git
